### ISSUE TITLE
280: Blender 2.80 support

### ISSUE THREAD

**Issue by dpospisil** [X-Plane/XPlane2Blender#399]
Since 2.8 beta is out, let's jump to 2.8!

Two reasons for switching to 2.8:
- no limit on number of collections. I think that 2.8 version exporter does not need to support two export strategies (root objects and layers), but just collections
- the 2.8 viewport is so much nice and responsive; it's a pleasure to work with

**Comment by tngreene** [comment]
I agree! So do some of the Laminar Research artists! But, there are some problems that will probably make this not happen for at least a year and a half if not more.

1. Blender 2.8 may have its Python API finished by now, but full release isn't scheduled until early 2019 (whatever that means). Most Blender users won't want to use a beta copy for production use.
2. Blender 2.8 represents a huge "no going back" moment, and the same would apply for XPlane2Blender. XPlane2Blender 2.79 would have to stop development of new features, and only the most critical of bugs would be developed for it. We'd have to make sure a very large percentage of artists are interested in switching before we say "upgrade or else!" or we'll lose people. In fact, some people are still using 2.49 because the cost of switching is so high they'd rather put up with it than commit. (Hence the converter!)
3. I'm guessing that many long term support distros won't upgrade either (*cough cough* Ubuntu *cough cough*). I've had someone say that they're stuck with 2.76 because their admin won't upgrade that software package. Some people couldn't switch even if they wanted to.
4. Root Objects mode and Layers Mode represent semantically different work flows, so we are keeping them no matter how use Blender's UI to say which one we're using. 2.49 had specially named objects, 2.7x has 3D view layers visible and the "Root Objects" checkbox. 2.8 will have....?

Blender doesn't care about how we use these things, so the burden to make it work and find a way to match the old style to the new without data lost or causing problems is entirely on us. That makes this potentially very difficult and certainly very risky. The risk makes me not want to do it above all else.

**So, perhaps in a year and a half when I hear developers chomping at the bit saying "We want 2.8!" we'll look at who is it it is available, how hard porting the plugin would be, how risky it would be, and how much time it would take and then decide.**

It certainly won't happen until after the converter is done.

If you'd like this to happen faster, I recommend evangelizing 2.8 as hard as you can and find a way to show the preferences of the community (including and especially payware artists who's business and food on the table literally depends on stable tools that don't cost them much time to switch). I tried making a poll once to ask simply "what version are you using" and got 25 votes, no replies hardly representative.

Simply "the UI is much better" isn't going to be enough of a selling point for most people. "The UI is much much much much better!" probably won't be either (see the fact that people are STILL using 2.49!)

Let me know what you find,
Ted

**Comment by dpospisil** [comment]
Many thanks for the detailed reply! Adding few comments.

1) Most Blender users won't want to use a beta copy for production use
That's clear. As a payware developer I would not want to switch to beta release either. However I would like to switch to upstream stable as soon as possible. Should the Blender 2.8 final release happen early 2019 it would be great if the exporter could follow soon.

2) XPlane2Blender 2.79 would have to stop development of new features, and only the most critical of bugs would be developed for it. 
Yes, users would need to migrate to 2.8 eventually. But I do not think that the cost of switching to 2.8 would be "so high". My experience is that switching to 2.8 is relatively easy and once you do that yo do not want to look back.

3) I'm guessing that many long term support distros won't upgrade either... Some people couldn't switch even if they wanted to.
This is very hard to buy as running blender is just a matter of downloading and unpacking a file. You do not need any admin privileges to do that.

4) "the UI is much better" isn't going to be enough of a selling point for most people
I did not want to say that the UI aesthetic is the selling point of 2.8. From my POV it is:

Viewport improvements 
- more options to visualise the model
- easy show/hide by type
- much much faster

Collections 
Collections are replacing layer system in <2.8. The huge advantage is that there is no limit on max number and they can also be nested. Currently we use exporting by layers mode and we need to divide the model into multiple scenes. Maintaining a model split into multiple scenes is a mess.

**Comment by tngreene** [comment]
Great points! And thank you for telling me about the issue with scenes. I don't honestly get to see a lot of "workflows". I only know about that when people say what feels good and bad.

**Comment by OzWookiee** [comment]
Being fairly new to Scenery and Aircraft (Helicopter) development learning Blender has been quite and journey and I've been watching the videos of the improvements made in 2.8 with tremendous interest as they really seem to have made workflows a priority and things certainly seem a LOT more intuitive and powerful.

I guess I'll have to experiment with 2.8 for personal growth and as you say, evangelise it so that we can get more interest in people switching.

**Comment by jamlen** [comment]
I am just starting out in my journey into Blender and have also found 2.80 much more straight forward to pick up and get going. I am also a software dev with python experience so would be happy to contrib to this end. @tngreene you mentioned the converter would need to be completed first... is there a project for the converter, as I can't find it!

**Comment by airfightergr** [comment]
@jamlen with 2.80 now in beta and very stable, it's the way to go for the near future, and 2.79 will soon become obsolete. Might end up in a situation like having the previous transition from 2.4 to 2.5. To have X2B scripts for the old version, and not the new.

If you can join in and help the migration to 2.80, will be pure awesomeness!

**Comment by tngreene** [comment]
@jamlen, there is a project. Its bugs and features that start with `2:49`. We also have a tag. #387 would be a good way to help the converter along that doesn't require much real coding. Researching how XPlane2Blender 2.49 used what Blender 2.49 data model states is rather time consuming as someone who never used it.

Even easier would be researching the various hard coded special strings that could have been used. See #381. Apparently naming an empty "OBJ" would make it into a root object. That isn't documented anywhere, but will be important to convert. If someone could research those things while I'm currently grinding through manipulators, that'd be a good place to start.

Just keep in mind, I have no idea what things are either.

**Comment by tngreene** [comment]
And again, the best thing for this bug report is to collect opinions and convince people it is the right move.

**Comment by agarzon** [comment]
I guess the "solution" for now is to model with blender 2.8 and make the export with 2.79.

Still, blender beta has been developed for a lot of months and is very stable at this point.

**Comment by OzWookiee** [comment]
I'm not sure 2.79 supports 2.8 files due to the changes with collections.

**Comment by agarzon** [comment]
I just confirmed it is not possible to port-back models made in 2.80 to 2.79.

A shame, modeling, texturing, sculpting on the new blender 2.80 is just awesome and extremely easy vs the 2.79 and previous versions. Not me saying it, there are thousands of 3D professionals reviews on forums and youtube that LOVE the new blender... still in beta however.

Looking forward to heard any 2.8 support in the upcoming years.

**Comment by aussig** [comment]
> I just confirmed it is not possible to port-back models made in 2.80 to 2.79.
You can, however, copy and paste between the two, although I'm having trouble getting UV mappings across for some reason.

**Comment by angelarchONE** [comment]
I just started with Blender after being a Maya vetran.  Blender 2.8 is magnificent and stable even in beta stage!  Blender 2.7 is utter garbage...  please oh please make an Xplane2Blender for 2.8 as soon as you can:)

**Comment by tngreene** [comment]
I have started a poll on x-plane.org. See [here](https://forums.x-plane.org/index.php?/forums/topic/176788-how-much-does-the-community-want-blender-28-support/). If you could answer it and *get as many other people to fill it out who aren't talking about it here* it would be very helpful. As I said before, getting a feel for how the wide community is feeling about it is very useful for determining how fast this happens.

The people who wouldn't sign up for github or don't know how matter for this decision. Maybe more active forum members can bump up the poll count and start a conversation.

**Comment by OzWookiee** [comment]
Blasted to all of my contacts!

**Comment by tngreene** [comment]
Future Transition resources:
- [2.8 Python API changes](https://wiki.blender.org/wiki/Reference/Release_Notes/2.80/Python_API)
- [2.8 Removed features](https://wiki.blender.org/wiki/Reference/Release_Notes/2.80/Removed_Features), 2.8 removes the game engine which is disastrous for converting 2.49 files. The new flow will have to be `Save with 2.49`->`Convert with XPlane2Blender for 2.79 (the final release we ever make for it)`->`Open in Blender 2.8`. Fortunately, I think this solution will continue work long enough for the community to save its old work. Otherwise, there are VMs, the old computers stuffed in the back of the closet, and all this stuff is free and open source.

**Comment by dpospisil** [comment]
I think that top level collection would be fine.
Or maybe let the user manually configure list of collections to export in the scene property tab?

**Comment by dpospisil** [comment]
This is still priority low? :(

**Comment by tngreene** [comment]
This doesn't change what I said: Probably nothing until summer.

**Comment by dpospisil** [comment]
Thanks, this looks much better now :)

**Comment by heclak** [comment]
Is someone already working on this or is there a dev branch to contribute to this issue? Would like to help speed this along if possible.

**Comment by tngreene** [comment]
Hi @heclak , as far as I know, no one is working on this, but, this is open source so if you'd like to contribute to making this go faster I would appreciate some help!

Its why I've been marking up the bug base with "help wanted" and "good first issue". The reason I didn't mark this one as "Good first issue" is because it touches massive portions of the code base and testing suite and it isn't a good first issue.

How well do you know XPlane2Blender, Python, Blender, and the unit testing suite?

**Comment by heclak** [comment]
Yea. I understand it’s a huge undertaking. I’ve been working on a conversion of an export plugin from 2.79 to 2.80 so I understand there’s a possible huge rewrite depending on how much legacy blender python api is in there. I’m rather familiar with working in Blender, modelling a lot but not much animation. Very comfortable with coding, python isn’t my primary language but I get by. Haven’t done unit testing in a long time. Last used was with JS. 

My problem with contributing would be my lack of familiarity of Xplane2Blender and the code base. I’ve only used it a long time ago for exporting some scenery objects. I will need to spend some time studying what the code is suppose to do first. 

I’m interested in contributing as it’s almost impossible for me to go back to 2.79. Totally spoiled by 2.80. 

Was asking if anyone is working on it offline first. Couldn’t find a working branch for it. There’s mention of a dev doc in the readme but I couldn’t find it. Is it still around?

**Comment by tngreene** [comment]
A dev doc in the read me? Can you give me a link?

I'm hesitant to start working on this this early for a couple of reasons:
1. I want to control this transition very very tightly, since the beginnings of projects have a great effect down the line. I also want an excuse to do some big clean up I haven't gotten a chance todo. As long as we're forced to rename a bunch of stuff, maybe it can be renamed properly!
2. Volunteer time isn't free. Even to update the docs of "where the important bits are" takes time, or if we were to jump on Skype and start talking, that is also time taken. I believe people are good until proven otherwise, but I don't have time to train up someone I don't know or trust someone with something so major or micromanage their work so its how I want. Again, no offense to you, honest truth. Me making a bunch of easy bugs marked "help wanted" was my hope of gaining some regular pull requests and talking with those people. No traction yet.
3. It is much easier to answer "When will 2.8 support happen?" with "Don't expect anything until July!" than answering "How come 2.8 broke my .blend file!" with "Use at your own peril. You are in peril and I cannot help you, only give you tools to shoot yourself in the foot with." People assume if it is public, it is supported. I don't want to put something out with absolutely no plans of supporting the people who use it. Even the pre-Alpha converter is getting support in the form of me explaining the risks to people personally and why something worked or didn't.
As soon as a whiff of working XPlane2Blender 2.8 code is working people will use it in droves.

4. My first priority is the converter. Currently I am supporting 2.49 files, 2.78 and 2.79 files. By making 2.8 code I am also supporting 2.8 (even if I say "no no guys, this is just a development branch! GitHub makes it all public wait stop aaarggkk **trampled**). I don't have the bandwidth for that and the 2.49 converter is *needed*, 2.8 is *wanted* (in my opinion because 2.79 is good enough for another month and a half).

In short, this needs to go right the first time, and I don't have much time for this or other developers I trust will know what to do without micromanagement. That could be you! I'm not trying to be a gatekeeper and a control freak, I'm just trying to wait until the time is right instead of making an inferior XPlane2Blender 2.8 and disappointing people or harming their work.

So, if you'd really like to help 2.8 happen faster cleaning up some of those little bugs I've marked as help wanted before 3.5.1 would be very helpful and a good way to learn XPlane2Blender's code base and get to know each other!

**Comment by tngreene** [comment]
Also, to clarify need vs want before I get bricks thrown at my head: Blender 2.49 is **literally** becoming unrunnable on people's computers, the work needs to get saved ASAP before people can't open their old stuff anymore. 2.8 does not even have a full release out yet (planned for July) meaning it shouldn't even be used on production stuff (see their recommendation and mine).

**Comment by agarzon** [comment]
Ladies and gentlemen... blender 2.8 has been officially released as the current stable version. 😄 2.79 is not anymore the latest version 🎆

Sorry for the excitement 😋

**Comment by arb65912** [comment]
I guess I have to say sorry for the excitement as well! :-)

**Comment by lehthanis** [comment]
How many of us have opened our projects up in 2.8 and switched to eevee for
the lulz. It's pretty stinkin amazing.

**Comment by tngreene** [comment]
Thanks for keeping me posted guys. Still working as hard as I can getting the converter to Alpha.1. Once we reach that, I'll stop working on the converter and work on this as feedback allows.

**Comment by tngreene** [comment]
So, good news, after much demand I figured I would work on the thing that more people have been asking for (2.8 Support). More good news, I have the UI back up. Fun times.

Bad News: Blender Render (aka Blender Internal) is removed, meaning the whole "Material has a Texture Slot Which has your Texture Image in it" Idea is not only dead, but on load, Image Datablocks aren't even linked to Anything anymore!

`The render engine Blender Internal was removed, replaced by EEVEE, the new real-time engine.`

There is nothing XPlane2Blender can do here, it happens on open in 2.8. New art will be fine, old art will have to choose between the "manually set texture" option and manually redoing the texturing assigning work for whatever we choose. This is extremely disappointing for backwards compatibility and development of XPlane2Blender 2.8.

I wish I could say here: "We can make a 2.8 alpha and only new work or dedicated people willing to retexture their model can use it", but it isn't that easy. XPlane2Blender has no clue about Nodes, and references to "texture_slots" are in many places across the code base.

To recap so far
1. Due to popular demand, I tried porting 2.8. Some API and renaming stuff went very well
2. I got confident to try opening a real example file and found that Blender Render was removed (what a thing to put in the footnotes of the release log Blender!!!), meaning all our "Texture Slot of Material Slot of Object" stuff is dead
3. XPlane2Blender is going to have to use EEVEE now, or something else
4. People are going to have to transition to using EEVEE or that something else before or after opening a file in 2.8

This means...... we get to make a design decision! What do we want the new workflow to be? What do we want EEVEE to mean to us?

Last time we got presented with this during the 2.49 to 2.79 we chose the "Texture Slot of the Material Slot of the Object." (Related #357 because how every Texture Slot in an OBJ has to be the same) I don't know how much of a success that re-use has been. What do you think? How do you use EEVEE?

This is a huge foundational decision for XPlane2Blender 2.8 so it won't be treated lightly.

**Comment by lehthanis** [comment]
Cycles is available in both 2.79 and 2.80 so it seems to be the least
common denominator. Also eevee uses the same nodes and shaders that cycles
uses. One of my art projects switched from cycles to eevee with zero issues
and shaders worked beautifully in realtime. My xplane projects all already
specify all textures in the root node so I won't be offended by this change
at all. I could be quite ignorant on the matter though.

**Comment by tngreene** [comment]
# Things To Do
- [ ] Keep going on the rest of Blender 2.8, force people to uncheck Autodetect Textures and override the paths if they want to continue
- [ ] Make a design decision on how to use Nodes
- [ ] If it is a terribly tedious solution, make a forwards compatibility script to run in 2.79 to make your projects Node Ready
- [ ] Replace * with @ for Vector and Matrix multiplication
- [ ] Replace old operators calls with new ones
- [ ] Keep using this sweet sweet opportunity for some major code clean up

# Design Questions?
- XPlane2Blender Property Workflow vs Node Workflow?
    Could we get more WYSIWYG behavior out of reusing Nodes and seeing it in real time?

Would that be too jarring for existing users to know which ones are special and re-used? What if the mapping is fishy?

What if it isn't really WYSIWYG because Blender and X-Plane don't render the same anyway?

- How are people currently using Nodes?

    Can we use this fact to shape our decision? Not step on anyone's toes and override parts of their workflow?

- What about (and this is just the devils advocate here) going back to 2.49's method of using the UV/Image editor's Image?

**Comment by tngreene** [comment]
> Cycles is available in both 2.79 and 2.80 so it seems to be the least
common denominator. Also eevee uses the same nodes and shaders that cycles
uses. One of my art projects switched from cycles to eevee with zero issues
and shaders worked beautifully in realtime. My xplane projects all already
specify all textures in the root node so I won't be offended by this change
at all. I could be quite ignorant on the matter though.

@lehthanis Can you please upload a file that demonstrates some Cube with a Texture Slot being turned into a Node based workflow to show your idea and how you do your projects? This is the sort of very valuable IRL workflow data I need

**Comment by lehthanis** [comment]
Absolutely. I'll have that for you before end of the day.

**Comment by lehthanis** [comment]
@tngreene  Ok, this file has a blender 2.79 file and a blender 2.80 file.  the 2.79 is a basic stylized fuselage with diffuse, nml, lit textures applied via nodes in cycles. This is a root objects type export, with the empty as the root.  The root does not autodetect textures and you see I have them listed there. The view I have set up is my custom node editing view using a principled BSDF shader with an emission shader added together to get the lighting. Not only will this render properly in cycles but it will export properly to x-plane. (Render will suck because there's no environment or exterior lighting)

The blender 2.80 file is the 2.79 file I opened in 2.80 and saved as new.  The only change is the render engine was set to eevee and I turned on some of the render features like AO and Bloom and SSR. As you can tell, out of the box, this looks MUCH better (and modeling can be done with this level of detail). but the node structure imported unmodified from cycles.  All of my airplane objects use this exact same node structure.

The only issue is I've never bothered with getting transparency/alpha to work in this shader config, but I can fool around with it if you like and provide updates.  But this should give you something to work with. 

[NodeWorkflow2.79-2.80.zip](https://github.com/X-Plane/XPlane2Blender/files/3462992/NodeWorkflow2.79-2.80.zip)

**Comment by lehthanis** [comment]
With this in mind. A 2.8 branch with no auto detect textures functionality would probably be immediately beneficial for people like me to get started testing the other aspects of the exporter. Again pardon my ignorance if it's not that simple.

**Comment by arb65912** [comment]
With all respect to other people opinions I have to say I have never had luck with "autodetect textures" feature and actually see it as a low priority.
Manually assigning png, lit.png and nml.png is a one time setup and does not take that long.

Personally the most important is that exporter properly exports collections to XP layers along with bone animations and regular animations as well.

All PBR happens in maps editing alpha and channels as long as NORMAL_METALNESS
GLOBAL_specular	1 are exported.

I am sorry if my comments might seem stupid to some but I am an average and happy 2.79 exporter user and I am sure many would share my opinion.

Thank you for the update Ted, greatly appreciated.

**Comment by tngreene** [comment]
@lehthanis and @arb65912 Neither of your comments are stupid or ignorant! This is really good feedback. I wish I had known that this feature was very under used. Referencing (#357) again and making the converter convert and apply all the GLOBAL_ attributes again I can see how much of a pain this "Same setting applied across all Materials/Textures is".

I still want to know more.

So, @lehthanis , all this Node stuff is just so you can see what the texture will look like on the plane? Where do you create and change the texture .pngs? Photoshop?

**Comment by lehthanis** [comment]
I imagine the majority of the modern graphics workflow is moving to substance painter. Mine is. I used to bake my normal maps and ambient occlusion in 3ds Max from masks and grey scale bump maps drawn in photoshop. But now I export a wavefront obj to substance painter after uv unwrapping and then export the textures from there using an export configuration I got from Ben Supnik which I uploaded to the .org (https://forums.x-plane.org/index.php?/files/file/43553-x-plane-substance-painter-export-presets/). I then bring them back into blender.. 

I haven't watched the videos all the way through but you should check out some jrolla YouTube videos. He does substance painter. I'm attaching a screenshot showing the result of placarding done in inkscape and imported as a mask in substance painter for both lit and color. I can even make the white paint have some small height and different specularity. Amazing. 

I'd love to participate in some workflow discussions with you and other artists. Perhaps we can get a common and repeatable workflow down. I imagine these pbr textures can be built up procedurally with blender shader nodes and masks and baked to texture in blender without using substance painter too. I just use substance painter because I can paint or mask these materials directly onto my mesh in there.

![Screenshot_2019-08-04_23-43-07_1920x1080](https://user-images.githubusercontent.com/4234630/62471831-7d1dcf00-b76b-11e9-8753-dd5521e2795b.png)

**Comment by tngreene** [comment]
@lehthanis Who is Jrolla? Can you link to the tutorial or video or page?

**Comment by lehthanis** [comment]
Sorry it's jrollon. Here's one of his substance videos. I think he uses my
gauge maker program in this too.

https://youtu.be/qM19ou0PfRU

**Comment by tngreene** [comment]
With #449 closed I've made my decision on the subject. Turns out it wasn't such a fundamental feature after all. Thanks all!

**Comment by Karl1206** [comment]
Hello Ted

Firstly thanks so much for the effort you are putting into this plugin. As a payware developer this is something so important to the day to day workflow. I think 2.8 is a huge step for Blender and although i am sure some people will stay with 2.79 for years to come i also think the opportunities and improvements of 2.8 will bring many new people to the platform, especially after the increased recognition in the industry. I really appreciate you already working on the plugin and i thought i would share a few of my thoughts especially on the material settings.

I would like to just say i am not a programmer so this is only what i as an artist would love to see meaning they might be easily be too complicated to program.

For the materials i think it would be fantastic if it could work with a node setup. It seems most programs are moving in the direction of nodes and i honestly find them really nice to work with. It would also allow for a preview method in either Eevee or cycles. This is increasingly important as we are moving more and more towards the PBR workflow which you can't properly visualise without some sort of render. Now with Eevee it would be a lot easier to preview it in real time. I would love to try to create a node group which could take the x-plane diffuse and normal map and show it in Eevee as accurately as possible.

Cheers, Karl

**Comment by tngreene** [comment]
@Karl1206 thanks for the response! This community involvement has been the best parts of the 2.8 transition project.

> I would love to try to create a node group which could take the x-plane diffuse and normal map and show it in Eevee as accurately as possible.

Is this something that XPlane2Blender needs to know about or a workflow artists can come up with on their own to make the textures referenced by the OBJ?

**Comment by Karl1206** [comment]
I was thinking something like what you can see in the picture I have attached here. Essentially a little node group included in the plugin where the user simply inputs the texture maps he is using and then it doubles as both a way for XPlane2Blender to see what textures should be used for what but also as a material setup for Eevee's renderer. 

I have already created this group but the only challenge right now is the calculation of the normal map's blue channel since X-Plane automatically calculates the blue channel's information in the simulator we need to do the same in Blender as the information isn't included. It should be possible to do as it is fairly simple math to calculate it but I haven't quite gotten it working yet. Anyways I hope this screenshot illustrates what I was thinking about a bit better.
![64209572fe24a4a4f24ba2ff71ef6451](https://user-images.githubusercontent.com/17119382/62736071-7b326500-ba24-11e9-8a21-e8d164375f2a.png)

**Comment by arb65912** [comment]
Where would be .lit texture in that diagram? Emissive?

**Comment by Karl1206** [comment]
@arb65912 Yes sorry I didn't expand the node quite all the way but yea the _Lit would be the Emissive. Of course, these names are just what I quickly gave them and since I work in Substance Painter all day I just used the names from that program.

**Comment by arb65912** [comment]
Thank you Karl! I will need to ask me texturing friend to give Substance Painter a try. 

I would like to hear your comments about it but I do not want to "pollute" that thread. Is there any way to contact you outside of here?

**Comment by tngreene** [comment]
Thank you for the picture. As talked about #449 in, a big question I asked is "How important is Autodetecting Textures if the OBJ spec only cares about a file path? How much effort is it worth avoiding simply filing in a textbox by hand?"

Something I admit I don't know a lot about, but from what other people have told me this is how it works. **Would you be able to use the above set up as a personal choice to generate the texture files, and tell XPlane2Blender the file path to those generated texture files?**

This type of suggestion feels more like a convention XPlane2Blender should pick up on rather than a necessary feature. The last convention picked didn't really get used by most apparently.

In case you haven't seen the manual override feature, aka unchecking Auto-detect Textures

![image](https://user-images.githubusercontent.com/5393207/62736805-45798600-b9fc-11e9-8e28-b7a6d2c0c77d.png)

**Comment by arb65912** [comment]
I would be fine with what is shown on above picture, meaning manual path typing.

**Comment by Karl1206** [comment]
Yea the manual texture selection is what I use on a day to day basis but this is mostly as I find the Blender Render material set up a bit confusing. I never really understood what tick boxes to tick to decide what textures should be used for what. My suggestion is definitely more of a nice to have than a need to have so don't blame you at all for not prioritizing it I just think it would be really cool if XPlane2Blender went more in the direction of nodes and integrated with Evee or Cycles for easy previewing. But regardless I will be super grateful for any way to export from 2.8 so whatever you end up doing I am happy :)

@arb65912 feel free to shoot me an email at karl120699@gmail.com :)

**Comment by tngreene** [comment]
Thanks for the input @Karl1206. I look forward to seeing what workflows people come up with. If there is anyway to give people XPlane2Blender Shader templates, that would be ideal. But first, lets get an exporter!

If anyone else has a reason we must absolutely come up with some convention for autodetection, e-mail me. Otherwise, I request we move on please.

</close conversation>

**Comment by BillAC** [comment]
I just wanted to chime in and thank @tngreene for his tremendous work on this project. Like many others, I cannot wait to start using Blender 2.8. Wish I knew Phyton a little better to be able to contribute. A word on backwards compatibility with older models: I'm not too worried since I can always fall back to older versions of Blender/scripts for the lesser used older models.

**Comment by dpospisil** [comment]
I also agree that texture auto detection is a minor issue I can live without. Many thanks @tngreene for the ongoing work!

**Comment by muchimi** [comment]
All - wanted to chime in as I'm a long-time Max user and now considering Blender 2.8 for some potential commercial work I want to do.  I think 2.8 has really a maturing of the platform and really holds its own against pillars like Maya and Max, sprinkled with a bit of Zbrush - while I know this transition will be quite difficult for me as old habits die hard, I think Blender with 2.8 is a very serious contender through my exploration so far.   The hold up now becomes: how to export to x-plane.

Really looking forward to export 2.8 scenes direct to x-plane.  Fine work on this plugin @tngreene.


Cheers

**Comment by tngreene** [comment]
I know things have been rather silent lately, but I wanted to share this very awesome picture. It features the AA gun from mobile showing its progress from 2.49 to 2.80. As you can see, the project has been preserved throughout the process perfectly. For this simple model the process took about 10 minutes and just worked (except the texture path I had to change. Use relative paths guys!).

Experimenting with taking the KingAir from 2.49->2.79->2.80 has also been very fruitful and has basically just worked also. It is obviously a much harder project to look and and declare it working.

This is a testament to Blender's backwards compatibility and the converter's level of care in conversion. Everything appears to be animating correctly also.
![aaa_gun_249_280_proof](https://user-images.githubusercontent.com/5393207/66161352-18abae80-e5fa-11e9-829d-0f4c299dfd68.png)

Root Objects + Manual Texture Paths should cover enough for people to get started with and be very happy I think.

Expect an alpha tomorrow or Monday so you can try this too!

**Comment by OzWookiee** [comment]
Oh wow that's pretty exciting! Much prefer the Root Objects method.

**Comment by muchimi** [comment]
Nice to hear!  I also prefer the root objects method - much more flexible.

**Comment by OzWookiee** [comment]
Is it worth Reviewing this setup to implement?
https://forums.x-plane.org/index.php?/files/file/51482-node-setup-for-x-plane-11-pbr-in-blender-28/&tab=comments

**Comment by muchimi** [comment]
Not sure - I do all my texturing in Substance - mostly because I know Painter and Designer pretty well.   I can't speak to the PBR workflow in Blender as I haven't moved that far.

**Comment by OzWookiee** [comment]
I imagine this workflow example is more about showing how those maps would look like in XP once exported from the PBR pipeline.  It could also allow others to create mats in Eevee that xp2b could understand and then export appropriately.

**Comment by tngreene** [comment]
# Things To Do

## General Stuff
- [x] Change Lamp to Light
- [x] Replace old operators calls with new ones
- [ ] Keep using this sweet sweet opportunity for some major code clean up
- [x] Replace * with @ for Vector and Matrix multiplication

## Layers Mode Replacement
- [x] See #450 

## LODS
- [x] See #451

## Materials and Textures
- [ ] Replace inspecting Texture Slot's File Paths with the Layer Settings's Texture File Paths, remove `Autodetect Textures` 
- [ ] Unit Tests that may need updating:
- features\texture_composition.test.py
- features\autodetect\autodetect_textures_2.test.py
- materials\objects\AOTC_2Mats_PNL.test.py
- materials\objects\COTC_2Mat_Hard.test.py
- materials\objects\IST_2Mats_SameSPEC_DifSURF.test.py
- materials\objects\IST_2Mats_wDraping.test.py
- materials\objects\SSO_2compatibleMats_Draped.test.py
- materials\objects\SSO_2incompatibleMats.test.py
- materials\objects\SSO_3compatibleMats.test.py

(The fact that the other **autodetect** unit tests didn't fail is very interesting to me!)
- [ ] Hide Autodetect Textures in the UI and delete it. If we come up with some new Autodetect system, we'll want people to opt into it
- [ ] If it is a terribly tedious solution, make a forwards compatibility script to run in 2.79 to make your projects Node Ready
- [ ] One day when we do have some conventions it would be nice to provide people templates of shading workflows. EEVEE is not instantly approachable.

# Blender 3D View
- [x] Pick your poison! EEVEE, UV/Image Editor preview, Substance Painter programs, or some other means I don't know about. Set the file path string to write to the OBJ and the rest is up to you on how you want to see it appear on your mesh as you author

**Comment by OzWookiee** [comment]
Finally getting back to Blender after a hiatus and have switched over to 2.8. my only issue now is that I can't find the obj textures references. I saw above some mention of the latter sounds but I was using the auto method and root object.

**Comment by tngreene** [comment]
@OzWookiee Blender 2.80 removed the whole "texture slots" concept altogether. Until we one day start using a Node Based autodetect solution, simply uncheck "Autodetect Textures" in the OBJ settings and fill in the file paths to where your texture files are manually.

![image](https://user-images.githubusercontent.com/5393207/72292646-cc700180-361f-11ea-92e0-e1b508849e93.png)

**Comment by tngreene** [comment]
With the release of beta.1 I'm calling this closed! Material stuff has been moved to #499! Thanks for all your help everyone! Looking forward to a good rc and beyond!
